stan/math/opencl/kernel_generator/binary_operation.hpp

```diff
@@ -0,0 +1,363 @@
+  using base::instance;
+  using base::var_name;
+
+  /**
+   * Constructor
+   * @param a first argument
+   * @param b sedond argument
+   * @param op operation
+   */
+  binary_operation(T_a&& a, T_b&& b, const std::string& op)  // NOLINT
+      : a_(std::forward<T_a>(a)), b_(std::forward<T_b>(b)), op_(op) {
+    const std::string function = "binary_operator" + op;
+    if (a.rows() != base::dynamic && b.rows() != base::dynamic) {
+      check_size_match(function.c_str(), "Rows of ", "a", a.rows(), "rows of ",
+                       "b", b.rows());
+    }
+    if (a.cols() != base::dynamic && b.cols() != base::dynamic) {
+      check_size_match(function.c_str(), "Columns of ", "a", a.cols(),
+                       "columns of ", "b", b.cols());
+    }
+  }
+
+  /**
+   * generates kernel code for this and nested expressions.
+   * @param[in,out] generated set of already generated operations
+   * @param ng name generator for this kernel
+   * @param i row index variable name
+   * @param j column index variable name
+   * @return part of kernel with code for this and nested expressions
+   */
+  inline kernel_parts generate(std::set<int>& generated, name_generator& ng,
+                               const std::string& i,
+                               const std::string& j) const {
+    if (generated.count(instance) == 0) {
+      kernel_parts a_parts = a_.generate(generated, ng, i, j);
+      kernel_parts b_parts = b_.generate(generated, ng, i, j);
+      generated.insert(instance);
+      var_name = ng.generate();
+      kernel_parts res;
+      res.body = a_parts.body + b_parts.body + type_str<ReturnScalar>::name
+                 + " " + var_name + " = " + a_.var_name + op_ + b_.var_name
+                 + ";\n";
+      res.args = a_parts.args + b_parts.args;
+      return res;
+    } else {
+      return {};
+    }
+  }
+
+  /**
+   * Sets kernel arguments for this and nested expressions.
+   * @param[in,out] generated set of expressions that already set their kernel
+   * arguments
+   * @param kernel kernel to set arguments on
+   * @param[in,out] arg_num consecutive number of the first argument to set.
+   * This is incremented for each argument set by this function.
+   */
+  inline void set_args(std::set<int>& generated, cl::Kernel& kernel,
+                       int& arg_num) const {
+    if (generated.count(instance) == 0) {
+      generated.insert(instance);
+      a_.set_args(generated, kernel, arg_num);
+      b_.set_args(generated, kernel, arg_num);
+    }
+  }
+
+  /**
+   * Adds event for any matrices used by this or nested expressions.
+   * @param e the event to add
+   */
+  inline void add_event(cl::Event& e) const {
+    a_.add_event(e);
+    b_.add_event(e);
+  }
+
+  /**
+   * Number of rows of a matrix that would be the result of evaluating this
+   * expression.
+   * @return number of rows
+   */
+  inline int rows() const {
+    int a_rows = a_.rows();
+    return a_rows == base::dynamic ? b_.rows() : a_rows;
```

Comment:
Than we return dynamic. An expression that has either rows or cols dynamic can not be evaluated. But it can be used in another expression. If it is, for example, added to a matrix (that has fixed rows and columns) the resulting expression will have dimmensions of the matrix.

Reply:
Scalars, for example, have dynamic rows and columns - they can be used in binary operations with a matrix of any size.